[tflite] expose more NNAPI Delegate to Java

freedomtan opened the pull request:
1. use StatefulNnApiDelegate instead of the deprecated `tflite::NnApiDelegate()`, see [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/nnapi/nnapi_delegate.h#L173-L181)
2. make it possible to specify NNAPI [execution preference](https://developer.android.com/ndk/reference/group/neural-networks#preferencecode) in Java.

freedomtan committed:
[tflite] expose more NNAPI Delegate to Java

1. use StatefulNnApiDelegate() instead of depecated one
2. make it possible to specify NNAPI execution preference in Java

jdduke commented on tensorflow/lite/delegates/nnapi/java/src/main/native/BUILD:
```diff
@@ -18,6 +18,8 @@ cc_library(
     ],
     deps = [
+        "//tensorflow/lite:context",
+        "//tensorflow/lite:framework",
```
Can we avoid pulling in the full framework here? Would like to avoid that dependency if possible.

jdduke commented on tensorflow/lite/delegates/nnapi/java/src/main/java/org/tensorflow/lite/nnapi/NnApiDelegate.java:
```diff
@@ -25,8 +25,74 @@
+  public static final class Options {
+    public Options() {}
+
+    /** undefined */
```
Maybe say "Undefined; specifies default behavior."

jdduke reviewed (commented):
Looks good, just a couple nits!

freedomtan committed:
address review questions

1. remove unnecessary dependency
2. more explanation for the default preference setting

jdduke commented on tensorflow/lite/delegates/nnapi/java/src/main/java/org/tensorflow/lite/nnapi/NnApiDelegate.java:
```diff
@@ -29,7 +29,9 @@
     public Options() {}
 
-    /** undefined */
+    /** undefined, specifies default behavior.
```
Nit: This needs a line break, fix to something like:

/** 
 * undefined, specifies default behavior.
 * So far, the default setting of NNAPI is EXECUTION_PREFERENCE_FAST_SINGLE_ANSWER.
  */

jdduke commented:
This is failing some internal tests, stay tuned.

jdduke commented on tensorflow/lite/delegates/nnapi/java/src/main/native/nnapi_delegate_jni.cc:
```diff
@@ -21,10 +21,28 @@ limitations under the License.
+  StatefulNnApiDelegate::Options options = StatefulNnApiDelegate::Options();
+  options.execution_preference =
+      (StatefulNnApiDelegate::Options::ExecutionPreference)preference;
+  options.accelerator_name = env->GetStringUTFChars(accelerator_name, NULL);
```
You need to make null checks here for the arguments, e.g.,

if (accelerator_name) {
  options.accelerator_name = env->GetStringUTFChars(accelerator_name, NULL);
}
...

freedomtan replied:
Thanks. Did that, please check the latest commit.

freedomtan replied:
Thanks. Did that, please check the latest commit.

freedomtan replied:
Thanks. Did that, please check the latest commit.

jdduke commented on tensorflow/lite/delegates/nnapi/java/src/main/native/nnapi_delegate_jni.cc:
```diff
@@ -21,10 +21,28 @@ limitations under the License.
+  options.cache_dir = env->GetStringUTFChars(cache_dir, NULL);
+  options.model_token = env->GetStringUTFChars(model_token, NULL);
+
+  return reinterpret_cast<jlong>(new StatefulNnApiDelegate(options));
```
You also need to free the UTFChars if they were created, e.g.,

std::unique_ptr<StatefulNnApiDelegate> delegate(new StatefulNnApiDelegate(options));
if (options.accelerator_name) {
  env->ReleaseStringUTFChars(accelerator_name, options.accelerator_name);
}

freedomtan committed:
add null checking and free

gbaned commented:
@freedomtan Could you please check reviewer comments and keep us posted. Thanks!

freedomtan commented:
@gbaned I did that already.
